core/Services/Calendar/Store.vala

```diff
@@ -1,307 +1,658 @@
             return client.is_readonly ();
         } else {
-            critical ("No calendar client was found");
+            critical ("No client was found for source '%s'", source.dup_display_name ());
         }
 
         return true;
     }
 
-    private async void add_event_async (E.Source source, ECal.Component event) {
-        unowned ICal.Component comp = event.get_icalcomponent ();
-        debug (@"Adding event '$(comp.get_uid())'");
-        ECal.Client? client;
-        lock (source_client) {
-            client = source_client.get (source.get_uid ());
+    public bool source_is_connected (E.Source source) {
+        return source_client.contains (source.get_uid ());
+    }
+
+    public void source_trash (E.Source source) {
+        sources_trashed.push_tail (source);
+        registry_source_removed (source);
+        source.set_enabled (false);
+    }
+
+    public void source_trash_undo () {
+        if (sources_trashed.is_empty ())
+            return;
+
+        var source = sources_trashed.pop_tail ();
+        source.set_enabled (true);
+        registry_source_added (source);
+    }
+
+    public void source_trash_empty () {
+        E.Source source = sources_trashed.pop_tail ();
+        while (source != null) {
+            source.remove.begin (null);
+            source = sources_trashed.pop_tail ();
         }
+    }
 
-        if (client != null) {
-            try {
-                string? uid;
-#if E_CAL_2_0
-                yield client.create_object (comp, ECal.OperationFlags.NONE, null, out uid);
-#else
-                yield client.create_object (comp, null, out uid);
-#endif
-                if (uid != null) {
-                    comp.set_uid (uid);
-                }
-            } catch (GLib.Error error) {
-                critical (error.message);
+    public E.Source? source_get_default () {
+        if (registry != null) {
+            switch (source_type) {
+            case ECal.ClientSourceType.EVENTS:
+                return registry.default_calendar;
+
+            case ECal.ClientSourceType.TASKS:
+                return registry.default_task_list;
             }
-        } else {
-            critical ("No calendar was found, event not added");
         }
+        return null;
     }
 
-    public void update_event (E.Source source, ECal.Component event, ECal.ObjModType mod_type) {
-        unowned ICal.Component comp = event.get_icalcomponent ();
-        debug (@"Updating event '$(comp.get_uid())' [mod_type=$(mod_type)]");
+    public void source_set_default (E.Source source) {
+        if (registry != null) {
+            switch (source_type) {
+            case ECal.ClientSourceType.EVENTS:
+                registry.default_calendar = source;
+                break;
 
+            case ECal.ClientSourceType.TASKS:
+                registry.default_task_list = source;
+                break;
+            }
+        }
+    }
+
+    public List<E.Source>? sources_list () {
```

Comment:
I don't think this needs to be nullable. According to valadoc, `list_sources` can't return null.

Reply:
This function _might_ be executed too early (when `registry` is not available yet), so it has nothing to return in this case. I have to admit this might happen very rarely. Are we fine with some weird exception in this case?

Reply:
Aha, got it. Then it makes sense to me as is.